packages/gatsby/src/internal-plugins/query-runner/query-runner.js

```diff
@@ -60,13 +60,32 @@ module.exports = async (pageOrLayout, component) => {
   const resultHash = md5(resultJSON)
   const resultPath = joinPath(
     program.directory,
-    `.cache`,
-    `json`,
+    `public`,
+    `static`,
+    `d`,
     pageOrLayout.jsonName
   )
 
   if (resultHashes[resultPath] !== resultHash) {
     resultHashes[resultPath] = resultHash
-    await fs.writeFile(resultPath, resultJSON)
+    const programType = program._[0]
+
+    // In production, write file to public/static/d/ folder.
+    if (programType === `build`) {
+      await fs.writeFile(resultPath, resultJSON)
+      return
+    }
+
+    // In development queue up results until a client is available
+    // push subsequent results to client
+    if (programType === `develop`) {
+      const result = {}
+      result[pageOrLayout.jsonName] = resultJSON
+      ws.pushResult(result)
+      const sockets = ws.instance()
+      if (sockets) {
+        sockets.emit(`queryResult`, result)
```

Comment:
This should only emit the `resultJSON` — for large sites, this could turn into megabytes of JSON.